[27기 강인웅][Mission 1~7 완료] [Mission 8 구현 중]위스타그램 클론코딩

wong0220 opened the pull request:
검색 기능 아직 미구현

## :: 최근 작업 주제 (하나 이상의 주제를 선택해주세요.)
- [x] 기능 추가
- [ ] 리뷰 반영
- [x] 리팩토링
- [ ] 버그 수정
- [ ] 컨벤션 수정

<br />

## :: 구현 목표 (해당 브랜치에서 구현하고자 하는 하나의 목표를 설정합니다.)
댓글 좋아요 기능 및 댓글삭제 기능
<br />

## :: 구현 사항 설명 (작업한 내용을 상세하게 기록합니다.)

1. 댓글 하트 클릭시 활성화,비활성화
- 댓글 작성으로 생성된 하트 이모티콘을 클릭시 함수가 발생한다
- 하트 이모티콘의 클래스는 3가지 클래스명이 들어 있는데 그 중 인덱스 2의 값 즉 3번째 클래스명으로 조건으로 구분해준다
- classList메소드의 2번째 인덱스가 'heart'일시 속이 채워진 하트 이모티콘으로 클래스 속성값을 변경한다.
- 또는 classList메소드의 2번째 인덱스가 'fullHeart'일시 속이 빈 하트 이모티콘으로 클래스 속성값을 변경한다

2. 댓글 삭제 기능 구현
- 댓글이 작성될 시 display:hidden 속성의 버튼을 하트 이모티콘 옆에 생성해준다.
- 댓글 div에 hover 시 display inline-block 속성값으로 버튼을 생성해준다
- 댓글 삭제 버튼 클릭시 해당 버튼 이벤트의 부모요소를 조상요소로부터 removechild 메소드로 제거해준다.

3.사용자 검색 기능 구현
- class를 이용해 5가지 예시의 사용자 인스턴스를 만들어 준 후 userList 배열에 넣어준다
- 검색창에 사용자를 검색시 기존의 사용자 객체가 담긴 배열을 forEach메소드를 통해 순환시켜준다
- 각 객체의 아이디에서 입력값의 길이만큼 substr을 뽑아준다.
- 뽑아준 문자열과 입력값의 문자열이 같다면 같은 사용자 객체만 동적으로 추천목록 박스에 넣어준다
- 추천되는 사용자의 데이터가 많아 박스의 max-height보다 길어진다면 자동으로 스크롤이 되어진다
- 지울 때도 마찬가지로 keyup될 때마다 해당하는 입력값으로 시작하는 사용자들을 추천해준다.
- 검색창에 문자를 입력할 때마다 추천되는 사용자의 수를 temp에 넣어준다
- 입력할 때나 지울 때 모두 인풋박스 하단에 생성되는 추천 박스의 자식들은 for문을 통해 temp만큼 반복해주며 지워준다.
<br />

## :: 성장 포인트 (해당 기능을 구현하며 고민했던 사항이나 새로 알게된 부분, 어려웠던 점 등을 작성합니다.)
- 처음 하트 이모티콘을 동적으로 추가해 줄 때 innerHTML메소드를 이용하여 기능을 구현해주었다. 하지만 보안상 약점이 있는 메소드란 것을 
알게됐고 클래스 속성과 속성값을 어떻게 줄까 고민하다가 setAttribute 메소드가 요소에 속성과 속성값들을 부여해 줄 수 있었고 이미 해당 요소에
같은 속성이 있을 경우 속성값을 새롭게 초기화 해준다는 것을 알게됐다. 이러한 특성으로 조건에 따라서 빈하트와 채운 하트를 구현할 수 있었다.

- 댓글 삭제 기능을 구현하려고 했을 때 처음에는 remove메소드를 이용하려고 했었다. 머릿속으로는 타겟의 조상요소에서 타겟의 부모요소를 제거해야겠다는 생각을 했었는데 remove와 removeChild 메소드의 차이를 잘 이해하지 못 해 동적으로 생성된 모든 div 태그들을 다 지워지는 일이 있었다. 알고보니 remove는 해당 노드를 지운다는 메소드이고 removeChild는 부모노드로부터 자식노드를 지우는 메소드였다. 나는 처음 내가 구상한 코드를 구현하기 위해 removeChild를 고수하였지만  현재 내 코드를 더 깔끔하고 짧은 코드로 만들려면 remove 메소드를 사용해서 구현하면 된다는 사실을 알게되었다.

- 검색 기능 구현을 하기 전 어떤 식으로 접근을 해야하나 고민을 많이 했고 /*
클래스로 유저 추천목록 생성
배열에 클래스로 생성한 인스턴스들을 담아줌
검색 인풋에 키 업 될 때마다 인스턴스가 담긴 배열 for each (el => {el['id'] === 인풋.value}) 후 recommend 변수에 담아줌
recommend변수에 담긴 배열 정보들을 검색창 밑에 동적으로 생성해 줌
*/ 이렇게 주석을 쓰며 크게 그림을 그려보았다.
이러한 그림을 바탕으로 작업하던 중 검색창에 문자를 칠 때마다 기존에 추천됐던 유저 목록이 계속해서 누적되는 현상이 발생했다. 이 문제를 해결하기 위해 추천목록을 감싸주는 컨테이너 박스를 삭제시키는 방법을 적용시켜봤지만 박스는 사라지더라도 동적으로 생성된 유저 데이터들은 컨테이너에 그대로 남아있었다. 그래서 키보드를 누를 때마다 추천되는 유저 수를 temp변수에 넣어주었고 그 temp의 수만큼 for문을 돌려 container박스에 계속해서 남아있는 유저 정보들을 부모.removeChild(부모.firstchild)를 통해 해결할 수 있었다. 또한 temp의 위치를 처음에 else문 안에 넣어 if문 조건을 만족했을 때 temp에 값이 할당 되지 않는 작은 실수를 해서 원인을 찾는데 꽤 오래걸렸고 결국 부모요소에 자식이 없는데 for문으로 자식을 없애라고 명령하니 오류가 났던 것을 알게되었다.

<br />

## :: 기타 질문 및 특이 사항
- 콜백함수 내의 클릭 이벤트 함수가 어떻게 실행되는 건지 직관적으로는 이해가 가지만 어떻게 실행될 수 있는 건지 잘 이해가 가지 않습니다
 (main.js 내의 90번째 줄과 91번째 줄)  댓글 작성 버튼을 클릭해서 생성이 되면 그 콜백함수 내의 또 다른 클릭함수가 실행되기 전까지는
댓글 작성 클릭 함수는 종료되지 않는 건가요?

Jun4928 commented on students/13th/soomyung/js/searchAlgo.js:
```diff
@@ -0,0 +1,137 @@
+//(e.g.) coming out "wecode" when searching "wcd", "cd", "oe", "ee" and so on...
+
+// search array creater: using subset Algo
+const searchArrayCreator = (str) => {
```
subset 알고리즘 까지 사용하셨네요!
코드 너무 멋집니다!!

VannsKang replied:
꼼꼼한 리뷰 감사합니당!
덕분에 좀더 나은 코드 된것 같아요 🙌🏻🙌🏻🙌🏻🙌🏻

VannsKang replied:
+ 다음 커밋부터는 컨벤션 지켜서 하도록 하겠습니당!

Jun4928 reviewed (commented):
수명님 정말 멋진 코드가 많네요!
고심한 흔적이 보입니다. 🔥👍
앞으로도 멋진 코드 기대할게요 😎

Jun4928 commented:
중간에 hidden conversations 모두 펴서 리뷰 확인 및 반영 해 주세요🙌

VannsKang commented on students/13th/soomyung/guide/events.md:
```diff
@@ -0,0 +1,10 @@
+
+# 메인 페이지
+1. 댓글 input 창에 엔터치거나 "게시" 누르면 댓글 추가되도록
+- createElement로 요소 생성해서, input에 입력한 값이 추가 되어야합니다.
```
주신 내용 참고해서 remote data delete 하고 gitignore 로 exception 설정해서 푸쉬했습니당~ 🤟🏻🤟🏻

VannsKang commented on students/13th/soomyung/guide/login.md:
```diff
@@ -0,0 +1,39 @@
+# 로그인 페이지
```
여기도용 🤟🏻

VannsKang commented on students/13th/soomyung/guide/main.md:
```diff
@@ -0,0 +1,47 @@
+# 메인 페이지
```
완료! 🤟🏻

VannsKang commented on students/13th/soomyung/js/login.js:
```diff
@@ -0,0 +1,33 @@
+const getPw = document.querySelector("#password");
+const getBtn = document.querySelector(".button");
+const getForm = document.querySelector(".form-section");
+const getError = document.querySelector(".error");
```
ID selector는 class selector로 치환하였고 나머지 동사형 변수들은 각자 명사형으로 정리했습니당 👍

VannsKang commented on students/13th/soomyung/js/login.js:
```diff
@@ -0,0 +1,33 @@
+    "<span>The username you entered doesn't belong to an account. Please check your username and try again.</span>";
+};
+
+// length check
```
삭제 완료🤟🏻

VannsKang commented on students/13th/soomyung/js/login.js:
```diff
@@ -0,0 +1,33 @@
+  verifyAccount(idIsTrue, pwIsTrue);
+});
+
+// check @ with Error Message
```
삭제 완료🤟🏻

VannsKang commented on students/13th/soomyung/js/login.js:
```diff
@@ -0,0 +1,33 @@
+  x && y
+    ? getBtn.classList.add("button-active")
+    : getBtn.classList.remove("button-active");
+};
```
argument renaming 완료! 🤟🏻

VannsKang commented on students/13th/soomyung/js/login.js:
```diff
@@ -0,0 +1,33 @@
+// length check
+getForm.addEventListener("keyup", (e) => {
+  getId.value.length >= 1 ? (idIsTrue = true) : (idIsTrue = false);
+  getPw.value.length >= 1 ? (pwIsTrue = true) : (pwIsTrue = false);
```
refactoring complete!
맨 밑에 형식으로 수정하였습니다! 훨 보기 좋네용 🤟🏻

VannsKang commented on students/13th/soomyung/js/login.js:
```diff
@@ -0,0 +1,33 @@
+  e.preventDefault();
+  getId.value.search("@") === -1
+    ? putErrorMessage()
+    : getError.removeChild(getError.childNodes[0]);
```
idInput.value.includes("@")
    ? errorOut.removeChild(errorOut.childNodes[0])
    : putErrorMessage();

다음과 같이 변환했습니당 🤟🏻

VannsKang commented on students/13th/soomyung/js/main.js:
```diff
@@ -0,0 +1,309 @@
+// for feed image responsive
+const getFdImg = document.querySelector(
+  ".feed-image>img[src='https://via.placeholder.com/612']"
+);
```
변수명 변경했습니당 👍

VannsKang commented on students/13th/soomyung/js/main.js:
```diff
@@ -0,0 +1,309 @@
+    id: "john_doe_1st",
+    name: "John Doe First",
+    profile: "https://via.placeholder.com/32",
+  },
+  {
+    id: "john_doe_2nd",
+    name: "John Doe Second",
+    profile: "https://via.placeholder.com/32",
+  },
+  {
+    id: "john_doe_3rd",
+    name: "John Doe Third",
+    profile: "https://via.placeholder.com/32",
+  },
+  {
+    id: "john_doe_4th",
+    name: "John Doe Forth",
+    profile: "https://via.placeholder.com/32",
+  },
+  {
+    id: "john_doe_5th",
+    name: "John Doe Fifth",
+    profile: "https://via.placeholder.com/32",
+  },
+  {
+    id: "john_doe_6th",
+    name: "John Doe Sixth",
+    profile: "https://via.placeholder.com/32",
+  },
+  {
+    id: "john_doe_7th",
+    name: "John Doe Seventh",
+    profile: "https://via.placeholder.com/32",
+  },
+  {
+    id: "john_doe_8th",
+    name: "John Doe Eighth",
+    profile: "https://via.placeholder.com/32",
+  },
+];
```
감사합니당 🙏🏻

VannsKang commented on students/13th/soomyung/js/main.js:
```diff
@@ -0,0 +1,309 @@
+  window.innerWidth < 612
+    ? (getFdImg.src = "https://via.placeholder.com/500")
+    : (getFdImg.src = "https://via.placeholder.com/612");
+});
```
감사합니당 🙏🏻 x 2

VannsKang commented on students/13th/soomyung/js/main.js:
```diff
@@ -0,0 +1,309 @@
+        `
+  );
+  getFriendUl.append(friendLi);
+};
```
감사합니당 🙏🏻 x 3

VannsKang commented on students/13th/soomyung/js/main.js:
```diff
@@ -0,0 +1,309 @@
+};
+
+// generator
+for (let i = 0; i < 5; i++) {
+  const friendListGenerator = () => {
+    // random name picker from the db
+    const randomIdx = Math.floor(Math.random() * userList.length);
+    const randomName = userList[randomIdx].id;
+    if (!friendCheck.includes(randomName)) {
+      // if there is no identical items in the friendcheck, push the item inside friendcheck array
+      friendCheck.push(randomName);
+      // generator
+      friendListElement(randomIdx);
+    } else if (friendCheck.length <= 5) {
+      // recurring the function for generating 5 rows
+      friendListGenerator();
+    }
+  };
+  // execute frinds generator
+  friendListGenerator();
+}
```
아하~ 전 5명 맞출려고 1번 돌때 중복체크하고 중복있으면 그 턴안에 돌릴려고 접근해보았는데 
알려주신 코드가 훨씬 더 간쥐가 나네염 🤟🏻🤟🏻🤟🏻
적용했습니당~
 
userList.forEach(() => {
  const friendListGenerator = () => {
    if (friendCheck.length >= 5) return;

    const randomIdx = Math.floor(Math.random() * userList.length);
    const randomName = userList[randomIdx].id;

    if (!friendCheck.includes(randomName)) {
      friendCheck.push(randomName);
      friendListElement(randomIdx);
    }
    friendListGenerator();
  };
  friendListGenerator();
});

VannsKang commented on students/13th/soomyung/js/main.js:
```diff
@@ -0,0 +1,309 @@
+  );
+  li.append(cmtTemplateGenerator());
+  // create new comment with condition
+  getCommentText.value !== ""
```
적용했습니당~ 🤟🏻🤟🏻

VannsKang commented on students/13th/soomyung/js/main.js:
```diff
@@ -0,0 +1,309 @@
+  // if there is input value, label placeholder remove
+  if (e.target.value !== "") {
+    getSearchLabelText.classList.add("display-none");
+  }
```
적용했습니당~ 🤟🏻🤟🏻🤟🏻🤟🏻

VannsKang commented on students/13th/soomyung/js/main.js:
```diff
@@ -0,0 +1,309 @@
+// search box additional action:
+// if click the body, disapear dropdownlist. then if back to click the search box again, reappear the list again too
+getSearchBox.addEventListener("click", () => {
+  if (getSearchBox.value != "") {
```
적용했습니당~ 🤟🏻🤟🏻🤟🏻🤟🏻🤟🏻🤟🏻

VannsKang commented on students/13th/soomyung/js/searchAlgo.js:
```diff
@@ -0,0 +1,137 @@
+          // matching userdata from original DB with the first element from the array matched with input value at least one
+
+          // then show the search result
+          if (user.id == conArray[0]) {
```
적용했습니당~ 🤟🏻🤟🏻🤟🏻🤟🏻🤟🏻🤟🏻🤟🏻🤟🏻

Commit:
위스타그램 클론코딩
검색 기능 아직 미구현

Commit:
Add: 검색 기능 추가
- 검색시 누를 때마다 해당 단어 혹은 문장으로 시작하는 계정을 추천해줌
- 현재는 class를 이용해 5명의 사용자 인스턴스를 넣어서 테스트 해봄 ex) wecode 계정

Modify:변수명 변경 검색창에 있는 돋보기 변경
- 알아보기 힘든 변수명과 함수명 그리고 주석들을 제거
- 기존의 돋보기 모양을 placeholder안에 넣어줌

Commit:
Add
- 기능추가에 따른 메인 css추가

Commit:
Modify
- 코드  리팩토링

Commit:
Add - 유저 프로필 창 구현 중
